### PR TITLE
Add context-changing function support

### PR DESCRIPTION
**Pull request by davepagurek**
Probably we should better handle cases when a `with` is a string or something, since I don't guard against that.

...and also some rubocop changes, because `%w()` starts getting to some perl-level unreadability and I think `!!` boolean casting is fine.

@AndrewMcBurney @icechen1

### CODE REVIEW

**Comment by davepagurek** [comment]
Oops my class comments for the new syntax nodes are copied and pasted, need to fix that

**Comment by armcburney** [comment]
I find it funny that the perl guy is advising us to not go with the perl-esque syntax ;)

File: src/ruby/Grammar.rb [hunk 1]
```diff
@@ -38,7 +40,7 @@ def self.print_failure(text)
     puts '===================================='
     text.each_line.with_index { |line, i| puts "#{i + 1} #{line}" }
     puts "====================================\n\n"
-    puts parser.failure_reason
+    puts @parser.failure_reason
```
Comment (armcburney) [inline src/ruby/Grammar.rb:43]: Why did you change this line? For consistency with the declaration?

File: src/ruby/PreProcessor.rb [hunk 2]
```diff
@@ -47,7 +47,7 @@ def process_emerald(input)
   def check_new_indent(new_indent)
     if new_indent > @current_indent
       new_indent_greater(new_indent)
-    elsif new_indent < @current_indent && new_indent.nonzero?
```
Comment (armcburney) [inline src/ruby/PreProcessor.rb:50]: Ahh yeah, this isn't necessary because I added `next if line.lstrip.empty?`. Thanks for catching that!

**Review by armcburney — APPROVED** [review]
Great work Dave!!

File: src/ruby/Grammar.rb (hunk 1 above)
Comment (davepagurek) [reply src/ruby/Grammar.rb:43]: Because `@parser` is the variable, `parser` is undefined. It wasn't printing anything before
